keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1004 @@
+
+* CRDs work well, but are inappropriate for use by core APIs.
+
+Each of these approaches would be simplified by an in-tree mechanism for
+distributing trust anchor sets.
+
+### Goals
+
+* Provide a Kubernetes data type that holds X.509 trust anchors for use wherever
+  a set of X.509 trust anchors is needed.
+
+* Make it easy for certificates/v1 signer implementations to publish their
+  relevant trust anchors using a TrustAnchorSets object.
+
+* Make it easy for workloads that need access to the trust anchors that back a
+  particular certificates/v1 signer to access them within the container
+  filesystem.
+
+* Allow cluster operators to configure webhook and aggregrated API server
+  backends using that use TrustAnchorSets instead of inline PEM data.
+
+### Non-Goals
+
+* Distribute a default set of "system" trust anchors using TrustAnchorSets.
+  (This is a natural extension, though).
+
+* Handle trust anchors expressed in other forms than PEM-wrapped, DER-formatted
+  X.509.
+
+## Proposal
+
+This proposal is centered around a new cluster-scoped TrustAnchorSets resource,
+initially in the certificates/v1alpha1 API group.  The TrustAnchorSets object
+can be thought of as a specialized configmap tailored to the X.509 trust anchor
+use case.  Introducing a dedicated type allows us to attach different RBAC
+policies to TrustAnchorSets, which will typically be wide-open for reading, but
+locked-down for writing.
+
+Each TrustAnchorSets object is a container for multiple named trust anchor sets.
+Most use cases will simply using a single trust anchor set named default, but
+mutual TLS use cases involving multiple federating meshes will require the
+ability to express multiple, non-commingled trust anchor sets.
+
+Like CertificateSigningRequest, TrustAnchorSets are scoped to a particular
+signer using a `signerName` field, which will also be usable in RBAC
+ClusterRoleBindings.  It is possible to grant specific users the ability
+read and/or update all TrustAnchorSets that correspond to a specific signer.
+
+To enable consumption by workloads, the new `trustAnchorSets` Kubelet projected
+volume source allows projecting a trust anchor set from a particular signer into
+the container filesystem, with the contents of the projected files updating as
+the corresponding TrustAnchorSets are updated.
+
+### User Stories
+
+#### Trust Anchor Distribution for Private CAs
+
+In my cluster, I operate a certificates/v1 signer implementation called
+`example.com/server-tls`.  This is backed by a private CA hierarchy.  By some
+unspecified mechanism, I issue certificates from this CA hierarchy to a server
+workload (a pod named `server/server`).  My client workload (a pod named
+`client/client`) needs access to the trust anchors for the CA in order to
+connect to the server.  The client should be able to continually establish
+connections to the server without downtime, even as I rotate the trust anchors
+that back the CA.
```

Comment:
But how do you know when all clients have read the updated CA?

Reply:
As described below, you basically have to do this out of band, by ensuring that clients reread the updated trust anchors file promptly.

A full end-to-end solution would involve each client exporting metrics indicating that it has picked up the new trust anchors, and not allowing the old trust anchors to be retired until you are satisfied with the uptake metrics.

Reply:
I would also be interested to read a more detailed use case of how CA rotation could be achieved with this mechanism as I think making it possible to rotate CAs easier would be very valuable.
Is there a way to observe when kubelet has updated all pods with the new CAs?
It feels like if it was the case that the client anyway needs to implement custom logic to export metrics about what CAs they use then it is actually not the case that this makes rotation much easier for clients (unless there is already some standard way for them to do that)

Reply:
Metrics from kubelet are a good idea.

They don't cover the end-to-end client experience, though.  When I've built a similar product, we basically went with "tell clients they need to notice changes within 5 minutes, and then wait several hours after each step in the CA rotation".  Workloads that fail to pick up the new changes will begin crashing or having connection errors, which can be handled by liveness / readiness probes.

Reply:
> I don't think that the kubelet has any way to know if the client application has started using the updated trust bundle.

Is kubelet able to observe reads on disk via the container runtime?  i.e. kubelet changes the bundle at time `T1` and observes a read of the file at time `T2` as a good (but still not perfect) indicator that the app is up to date?

Reply:
I don't think so, unless we do something weird like implement projected volumes in terms of FUSE.